JENKINS-40300: Add MemoryReservation to set soft memory limits on the containers

nullify005 opened the pull request:
Provide support for setting of the soft memory (memoryReservation) and or the hard memory (memory) settings when creating the task container.

Even though the documentation for the API specifies that it's possible to not set either of these, AWS prevents creation of task templates which do not have one of either memory or memoryReservation set.

By adding the soft memory setting it means that you can have a container spawn & use more than the memory reservation.

In my context setting the hard limit was leading to the kernel oom-killer process often killing the Jenkins Slave Agent JVM & I'd much prefer this not to happen.

The soft limit is a cue to the scheduler to not overcommit memory, so it's fair enough that's it's there

ukayani commented:
We've run into similar issues with our slave due to the hard memory limit. Would love to see this change incorporated.

nullify005 commented:
Hi @roehrijn, I'll check it out & revise the PR :)

nullify005 commented:
Ok, bear with me because I've never done any form validation in Jenkins before.

* moved withMemoryReservation out as a default setting and only apply it to the task def if it's > 0 (should correct the issue you saw)
* added form validation for memory & memoryReservation fields to assist in proper settings
* renamed the fields the the jelly config to be more obvious (the naming in AWS could be confusing)
* added description fields for memory & memoryReservation

One thing I noticed is that even though the validation fails it can still try to apply / save, I'd like to inhibit saving if the fields aren't valid but I don't know Jenkins plugin / UI coding enough to know how to achieve that.
